onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -33,12 +33,14 @@ onnxruntime::NodeArg* AddCastNode(onnxruntime::Graph& graph,
                                   TypeProto* new_type,
                                   bool new_on_input,
                                   int64_t to_type,
-                                  onnxruntime::ProviderType providerType) {
+                                  onnxruntime::ProviderType providerType,
+                                  const char* insertCastPrefix) {
   //insert cast op to cast input
   int id = id_generator.Next();
 
   char str[32];
-  snprintf(str, 32, "CastDef_%d", id);
```

Comment:
Would be more standard to use Graph::GenerateNodeName() and use the return value in the AddNode call for the node name. Not sure we need to change the NodeArg name logic.

Reply:
Make sense. But I found a way to logically detect whether the node has been cast before without considering the name so this modification is deprecated. Thanks.

Reply:
Sorry - not quite following but maybe I don't understand the issue fully. I thought there was a conflict due to the naming being hardcoded. I perhaps misinterpreted that as being around the node name, but on re-reading the issue it seems like it's about the output name. Either way, those issues potentially apply to any node/value in the graph and aren't limited to Cast nodes. 

Based on that, simply changing to use Graph::GenerateNodeArgName and GenerateNodeName would address the issue with no other changes required. Less code, avoids conflict with anything in the graph and not just Cast nodes.

We should be adding a unit test for this as well to validate the change and prevent any future regressions.



---
In reply to: [471594481](https://github.com/microsoft/onnxruntime/pull/4811#discussion_r471594481) [](ancestors = 471594481)

Reply:
Sorry that I did not make myself clear.
I did try to solve it by creating new Node name instead of duplicated name, but it will keep adding new duplicated Cast node if the user keeps `InferenceSession` the produced model. So I suppose the check is needed. 

Still, I think changing to `Graph::GenerateNodeArgName` and `GenerateNodeName` can prevent similar errors in the future.
I will do it as well and add a unit test. Thank you for the suggestion @skottmckay

Reply:
It seems like the real issue is adding a Cast node when it's not needed - specifically to an existing Cast node that has a float16 input. 

To avoid that, ForceSingleNodeCPUFloat16ToFloat32 could ignore Cast nodes with the below change in the 'if' condition prior to unsetting the EP type. That results in NeedInsertCast not firing for Cast nodes (as they will still be assigned to an EP). With just that change everything seems to be happy (although we should use the name generation functions from Graph). 

```
Status ForceSingleNodeCPUFloat16ToFloat32(onnxruntime::Graph& graph) {
  // if graph only contain 1 compute node, don't force to float32
  if (graph.NumberOfNodes() <= 1) {
    return Status::OK();
  }

  for (auto& node : graph.Nodes()) {
    if (node.OpType() != "Cast" && IsSingleInputNodeFloat16Node(node)) {
      node.SetExecutionProviderType("");
    }
  }

  return Status::OK();
}
```

---
In reply to: [472419982](https://github.com/microsoft/onnxruntime/pull/4811#discussion_r472419982) [](ancestors = 472419982)

Reply:
Nice suggestion. It makes this modification much more beautiful and compact. Besides `node.OpType() != "Cast"`, do you think we need to additionally check the `to` attribute (equals to 1 -> convert to `float32`)? I am considering if the  `to` attribute in the Cast Node is a type other than `float32`, don't we need to insert a Cast node?

Reply:
I'm not sure we need any of the other changes like hasNodeBeenCast. The model from the github issue works fine just by avoiding adding another Cast to an existing Cast node. 

I don't think a check on the 'to' is required. You have 2 cases. A float16 going into a non-Cast node (need to insert Cast), or a float16 going into a Cast node. Even if that Cast node is to something other than float, inserting another Cast just goes float16 -> float -> output type of existing Cast node, so it doesn't change the end result.